fuzzamoto-ir/src/generators/block.rs

```diff
@@ -27,72 +128,90 @@ impl<R: RngCore> Generator<R> for BlockGenerator {
-            .force_append_expect_output(vec![begin_txs_var.index], Operation::EndBlockTransactions);
+        Ok(())
+    }
 
-        let block_version_var =
-            builder.force_append_expect_output(vec![], Operation::LoadBlockVersion(5));
+    fn name(&self) -> &'static str {
+        "BlockGenerator"
+    }
+}
 
-        let coinbase_tx_var =
-            if let Some(coinbase_var) = builder.get_random_variable(rng, Variable::CoinbaseTx) {
-                coinbase_var
-            } else {
-                self.coinbase_generator.generate(builder, rng, meta)?;
-                builder
-                    .get_random_variable(rng, Variable::CoinbaseTx)
-                    .unwrap()
-            };
+impl Default for BlockGenerator {
+    fn default() -> Self {
+        Self {
+            coinbase_generator: CoinbaseTxGenerator::default(),
+        }
+    }
+}
 
-        let block_and_header_var = builder.force_append(
-            vec![
-                coinbase_tx_var.index,
-                header_var.index,
-                time_var.index,
-                block_version_var.index,
-                end_txs_var.index,
-            ],
-            Operation::BuildBlock,
-        );
+/// `TipBlockGenerator` generates instructions for creating a new block on top of the current tip.
+pub struct TipBlockGenerator {
+    coinbase_generator: CoinbaseTxGenerator,
+    headers: HashMap<BlockHash, Header>,
+}
 
-        let conn_var = builder.get_or_create_random_connection(rng);
-        builder.force_append(
-            vec![conn_var.index, block_and_header_var[0].index],
-            Operation::SendHeader,
-        );
-        builder.force_append(
-            vec![conn_var.index, block_and_header_var[1].index],
-            Operation::SendBlock,
-        );
-        builder.force_append(
-            vec![block_and_header_var[2].index],
-            Operation::TakeCoinbaseTxo,
-        );
+impl<R: RngCore> Generator<R> for TipBlockGenerator {
+    fn generate(
+        &self,
+        builder: &mut ProgramBuilder,
+        rng: &mut R,
+        meta: Option<&PerTestcaseMetadata>,
+    ) -> GeneratorResult {
+        let Some(header_var) = select_header_nth(&self.headers, builder, meta, 0) else {
+            return Ok(());
+        };
 
+        let (_header, _block) =
+            build_block_from_header(&self.coinbase_generator, builder, rng, header_var, meta)?;
         Ok(())
     }
 
     fn name(&self) -> &'static str {
-        "BlockGenerator"
+        "TipBlockGenerator"
+    }
+
+    fn choose_index(
+        &self,
+        program: &crate::Program,
+        rng: &mut R,
+        meta: Option<&PerTestcaseMetadata>,
+    ) -> Option<usize> {
+        if let Some(meta) = meta.as_ref()
+            && let Some(nth) = meta.recent_blocks.iter().rev().nth(0)
+            && let Some((_, inst)) = nth.defining_block
+        {
+            let from: usize = inst + 1;
+            program.get_random_instruction_index_from(
```

Comment:
I fixed a bug here; previously this line was Some(inst + 1);

previously, I inserted the instructions to create a second block right after the BuildBlock operation of the first block.
But this would cause problems as this means that we create and send the 2nd block even before sending the first block.

instead, I changed it to select a random insertion point as long as it results in a valid program.
Maybe we could record `SendBlock`'s position. and hardcode the position, but for a fuzzer, selecting a random position would be more interesting (?) as it can sometimes generate a block whose previous block is not sent yet.